### PR TITLE
Feature: Support calculation of multi-k point band-decomposed charge density under LCAO basis

### LINKED ISSUE #2905: Band decomposed charge densities of any K point
### Background

At present, it seems that ABACUS can only calculate the partial (band decomposed) charge density for $\Gamma$ point according to parameter  `calculation = get_pchg`.

### Describe the solution you'd like

It is suggested to increase the function for band decomposed charge densities of **any K point**.

### Additional Context

In addition, I hope to get some help from the developers on how to calculate the partial charge density when the atomic orbitals and coefficients are obtained. 

I 'm developing post-processing functionality for DeepH programs, but I need some help in theory and technology, and I 'll be grateful !

### LINKED ISSUE #3572: Feature: Support calculation of multi-k point band-decomposed charge density under LCAO basis.
### Background

Currently, ABACUS only supports multi-k point calculation of band-decomposed charge density under PW basis, while under LCAO basis the calculation can be carried out only when `gamma_only` is set to be true. It is necessary to support calculation of multi-k point charge density under LCAO basis when INPUT keyword is set as `calculation = get_pchg`.

### Describe the solution you'd like

Changes of INPUT keywords:
1. Delete INPUT keyword `nbands_istate`.
2. Add a new INPUT parameter `bands_to_print`(similar to `ocp_set`), which enables users to choose which band to calculate.
3. Add a new INPUT boolean parameter `out_all_kpt`, which enables users to choose whether to calculate all k-points or gamma point only.

Possible changes of source code: `source/module_io/istate_charge.cpp`. Currently the calculation of density matrix under LCAO basis set is still using old-fashioned data structure. We plan to rewrite these codes with new density matrix using HContainer (see `source/module_elecstate/module_dm/density_matrix.cpp`).

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

Only source code in `module_io` will be changed. See solution description.

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

For a more detailed description (formulae etc.), see Feishu doc [https://xmywuqhxb0.feishu.cn/docx/FBHldIvHxoP99axSZeLcHIenntg](url). One can ask @AsTonyshment for access permission.

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### LINKED ISSUE #4412: Need to refactor `istate_envelope.cpp` and add INPUT parameter `bands_to_print` for `get_wf` calculation
### Background

As is stated in the title, we only have INPUT parameter `bands_to_print` for `get_pchg` calculation. It is necessary to implement this param in `get_wf` calculation as well. Furthermore, this is also a relatively independent part of `LOWF` refactoring plan.

### Describe the solution you'd like

I will implement this.

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### LINKED ISSUE #4697: Partial charge for spin down in current version appears to be zero for all bands
### Describe the bug

When setting `calculation = get_pchg` and `nspin = 2`, it appears that the spin-down component is not being calculated correctly. This results in the partial charge for spin down being zero across all bands. Furthermore, the cube file name of spin-down component is not correctly set, either.

I will work on a fix for this issue.

### Expected behavior

_No response_

### To Reproduce

_No response_

### Environment

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

### PR DESCRIPTION
**Pull request by AsTonyshment**
### Linked Issue
Fix #2905, fix #3572, fix #4412, fix #4697.

### What's changed?
- Added support for calculating band-decomposed charge density for multiple $\boldsymbol{k}$-points.
- Refactored `istate_charge.cpp` by extracting large segments of reused code into separate functions.
- Fixed the issue where charge density calculation was incorrect when `nspin=2`.
- Corrected the naming of output `.cube` files for the spin-down component when `nspin=2`.
- Standardized the $\boldsymbol{k}$-point and spin indices in output file names to start from 1.
- Fixed the issue where `bands_to_print` was not working properly due to recent INPUT refactoring.
- Added a new INPUT parameter `if_separate_k` to specify whether to write the partial charge densities for all $\boldsymbol{k}$-points to individual files or merge them.
- Updated the online documentation for ABACUS.

### Any changes of core modules? (ignore if not applicable)
- Added a new `DensityMatrix` function to obtain `DMR` for certain $\boldsymbol{k}$-point.

### CODE REVIEW

**Comment by AsTonyshment** [comment]
@Qianruipku Maybe you can have a look at whether I used the new INPUT formalism correctly? I added a system parameter `out_band_kb`, which is obtained by parsing the direct input `bands_to_print`.

File: source/module_io/bcast_globalv.cpp [hunk 1]
```diff
@@ -18,5 +18,9 @@ void ReadInput::set_globalv_bcast()
     add_bool_bcast(sys.double_grid);
     add_double_bcast(sys.uramping);
     add_string_bcast(sys.global_calculation);
+
+    // We need to broadcast the size of out_band_kb first, then broadcast the content of out_band_kb
```
Comment (Qianruipku) [inline source/module_io/bcast_globalv.cpp]: This annotation can be deleted. add_*vec_bcast now does not have this limitation.

File: source/module_io/input_conv.cpp [hunk 2]
```diff
@@ -43,90 +43,9 @@
...
-            T occ;
-            convert >> occ;
-            vec.emplace_back(occ);
-        }
-        regfree(&sub_reg);
-    }
-    regfree(&reg);
-}
+// Explicitly instantiate the template function
+template void Input_Conv::parse_expression<int>(const std::string& fn, std::vector<int>& arr);
```
Comment (Qianruipku) [inline source/module_io/input_conv.cpp]: These can be deleted. Because parse_expression is moved to *.h file.

**Review by Qianruipku — APPROVED** [review]

**Review by Qianruipku — APPROVED** [review]

File: source/module_io/istate_charge.cpp [hunk 3]
```diff
@@ -281,13 +163,330 @@ void IState_Charge::begin(Gint_Gamma& gg,
...
+                              << ik + 1 << "..." << std::endl;
+
+                    DM.init_DMR(GridD_in, ucell_in);
+                    DM.cal_DMR(ik);
+                    gk.initialize_pvpR(*ucell_in, GridD_in);
+                    gk.transfer_DM2DtoGrid(DM.get_DMR_vector());
+                    Gint_inout inout(rho, Gint_Tools::job_type::rho);
+                    gk.cal_gint(&inout);
+
+                    double** rho_save = new double*[nspin]; // Initialize an array of pointers
```
Comment (mohanchen) [inline source/module_io/istate_charge.cpp]: I would suggest using std::vector instead of new and delete

File: source/module_io/istate_charge.cpp [hunk 4]
```diff
@@ -281,13 +163,330 @@ void IState_Charge::begin(Gint_Gamma& gg,
...
+            if (value != 0 && value != 1)
+            {
+                ModuleBase::WARNING_QUIT(
+                    "IState_Charge::select_bands",
+                    "The elements of `bands_to_print` must be either 0 or 1. Invalid values found!");
+            }
+        }
+        // Fill bands_picked_ with values from out_band_kb
+        // Remaining bands are already set to 0
+        int length = std::min(static_cast<int>(out_band_kb.size()), nbands);
```
Comment (mohanchen) [inline source/module_io/istate_charge.cpp]: add const before int if applicable

File: source/module_io/istate_charge.cpp [hunk 5]
```diff
@@ -313,10 +514,51 @@ void IState_Charge::idmatrix(const int& ib,
...
+                             const double& nelec,
+                             const int nlocal,
+                             const ModuleBase::matrix& wg,
+                             elecstate::DensityMatrix<std::complex<double>, double>& DM,
+                             const K_Vectors& kv)
+{
+    ModuleBase::TITLE("IState_Charge", "idmatrix");
+    assert(wg.nr == kv.get_nks());
+
+    int fermi_band = static_cast<int>((nelec + 1) / 2 + 1.0e-8);
```
Comment (mohanchen) [inline source/module_io/istate_charge.cpp]: add const before int?

**Review by mohanchen — APPROVED** [review]

**Comment by mohanchen** [comment]
Why the docs/readthedocs.org:abacus-rtd has stopped? I think it doesn't related to the codes.

**Pull request merged**